enterprise/kernel/src/main/java/org/neo4j/kernel/impl/store/format/busted/BaseBustedRecordFormat.java

```diff
@@ -0,0 +1,244 @@
+import static org.neo4j.kernel.impl.store.RecordPageLocationCalculator.pageIdForRecord;
+import static org.neo4j.kernel.impl.store.format.busted.Reference.PAGE_CURSOR_ADAPTER;
+
+/**
+ * Base class for record format which utilizes dynamically sized references to other record IDs and with ability
+ * to use record units, meaning that a record may span two physical records in the store. This to keep store size
+ * low and only have records that have big references occupy double amount of space. This format supports up to
+ * 58-bit IDs, which is roughly 280 quadrillion. With that size the ID limits can be considered busted,
+ * hence the format name. The IDs take up between 3-8B depending on the size of the ID where relative ID
+ * references are used as often as possible. See {@link Reference}.
+ *
+ * For consistency, all formats have a one-byte header specifying:
+ *
+ * <ol>
+ * <li>0x1: inUse [0=unused, 1=used]</li>
+ * <li>0x2: record unit [0=single record, 1=multiple records]</li>
+ * <li>0x4: record unit type [0=first, 1=consecutive]
+ * <li>0x8 - 0x80 other flags for this record specific to each type</li>
+ * </ol>
+ *
+ * NOTE to the rest of the flags is that a good use of them is to denote whether or not an ID reference is
+ * null (-1) as to save 3B (smallest compressed size) by not writing a reference at all.
+ *
+ * For records that are the first out of multiple record units, then immediately following the header byte is
+ * the reference (3-8B) to the secondary ID. After that the "statically sized" data and in the end the
+ * dynamically sized data. The general thinking is that the break-off into the secondary record will happen in
+ * the sequence of dynamically sized references and this will allow for crossing the record boundary
+ * in between, but even in the middle of, references quite easily since the {@link DataAdapter}
+ * works on byte-per-byte data.
+ *
+ * Assigning secondary record unit IDs is done outside of this format implementation, it is just assumed
+ * that records that gets {@link #write(AbstractBaseRecord, PageCursor, int, PagedFile) written} have already
+ * been assigned all required such data.
+ *
+ * Usually each records are written and read atomically, so this format requires additional logic to be able to
+ * write and read multiple records together atomically. For writing then currently this is guarded by
+ * higher level entity write locks and so the {@link PageCursor} can simply move from the first on to the second
+ * record and continue writing. For reading, which is optimistic and may require retry, one additional
+ * {@link PageCursor} needs to be acquired over the second record, checking {@link PageCursor#shouldRetry()}
+ * on both and potentially re-reading the second or both until a consistent read was had.
+ *
+ * @param <RECORD> type of {@link AbstractBaseRecord}
+ */
+public abstract class BaseBustedRecordFormat<RECORD extends AbstractBaseRecord>
```

Comment:
This class is public abstract, but all abstract methods are protected, so modifiers need to be aligned